### ISSUE TITLE
MockServletContext should treat InvalidPathException like an IOException

### ISSUE THREAD

**Issue by Spunc** [spring-projects/spring-framework#23717]
I have a simple `ConfigurationProperties` class that binds a String property to a `Path` object:

```java
@Component
@ConfigurationProperties("app")
@Getter
@Setter
public class AppConfigurationProperties {
	
	private Path somePath;
}
```

Then I define `app.somePath=C:\\temp` within _application.properties_. This works when running the application, but a simple Test that loads the context throws:

```
com.example.demo.DemoApplicationTests > contextLoads() FAILED
    java.lang.IllegalStateException
        Caused by: org.springframework.boot.context.properties.ConfigurationPropertiesBindException
            Caused by: org.springframework.boot.context.properties.bind.BindException
                Caused by: org.springframework.core.convert.ConverterNotFoundException
```

The test class is most simple:

```java
@SpringBootTest
public class DemoApplicationTests {

	@Test
	public void contextLoads() {
	}
}
```

I created a project with the most recent Spring-Boot version (2.1.8.RELEASE) with Initializr (Gradle build with Spring Web as only dependency) to replicate the problem in isolation. It can be found [here](https://github.com/Spunc/demo). I run the project on Windows 10.

I found out that the test passes, if there are no backslashes in the property value. Furthermore, it passes if I don't use `org.springframework.boot:spring-boot-starter-web` but `org.springframework.boot:spring-boot-starter` as dependency. (The test project does not require web, but the actual project does.)

**Comment by Spunc** [comment]
Thank you very much for your replies. At first, I was surprised that @mbhave could not reproduce the problem. I had colleagues verify the issue and indeed, those with Windows have the problem while those with Mac-OS are not affected.

Thanks, @wilkinsona for your suggestion of a workaround. We will apply that.

**Comment by sbrannen** [comment]
Thanks for the great detective work, @wilkinsona!

We'll look into it.

**Comment by xak2000** [comment]
I also run into this problem with spring 5.1.5. Now I updated to 5.1.11 and the problem "solved", but now it prints warning to the log.

@sbrannen @wilkinsona 
I don't understand why `org.springframework.mock.web.MockServletContext.getResourceLocation(String path)` should prepend `resourceBasePath` to an absolute Windows path like `C:\temp`? So the result path is `src/main/webapp/C:/temp`. This would never work anyway.

If we leave `/C:/temp` here, then `this.file.toPath()` would not throw. It creates `WindowsPath("C:\temp")` instance.

Maybe this is incorrect to leave it like this because this is servlet context anyway (I'm not sure here what's the intention). But it is also have no sense to just do a concatenation like this. Maybe it works on unix systems, where any absolute path can be converted to relative just by appending another path at the beginning, but it doesn't work on Windows, where path can contain a hard drive letter.

If the path must really be prepended with servlet context path here, maybe it is better to remove the letter from `path` before doing this? So `/C:/temp` will become `/temp` and then `src/main/webapp/temp`. I can't reason about this because to be honest I don't fully understand this code and why we need to manipulate paths from `application.properties` this way in tests. If I created a property like `my-location=C:\temp` and run test with it, I don't expect it to be manipulated in some ways to be changed to point to some another path in a servlet context. Maybe for relative paths, but not for absolute.
